snapcraft_parser/main.py

```diff
@@ -0,0 +1,230 @@
+        origin_type = data[key].get("origin-type")
+        project_part = data[key].get("project-part")
+        subparts = data[key].get("parts", [])
+
+        if origin is not None:
+            # TODO: this should really be based on the origin uri not
+            # the part name to avoid the situation where there are multiple
+            # parts being pulled from the same repo branch.
+            origin_dir = os.path.join(BASE_DIR, key)
+            try:
+                os.makedirs(origin_dir)
+            except FileExistsError:
+                pass
+
+            class Options():
+                pass
+
+            options = Options()
+            setattr(options, "source", origin)
+
+            if origin_type is not None:
+                setattr(options, "source_type", origin_type)
+
+            sources.get(origin_dir, None, options)
+
+            origin_data = _get_origin_data(origin_dir)
+
+            parts = origin_data.get('parts', {})
+
+            source_part = parts.get(project_part)
+            if source_part is not None:
+                after = source_part.get('after', [])
+
+                if after:
+                    after = _update_after_parts(project_part, after)
+                    after_parts.update(set(after))
+
+                parts_list[project_part] = source_part
+                subparts_list, subparts_after_list = _process_subparts(
+                    project_part, subparts, parts)
+                parts_list.update(subparts_list)
+                after_parts.update(subparts_after_list)
+
+        if is_valid_parts_list(parts_list, after_parts):
+            master_parts_list.update(parts_list)
+
+        return master_parts_list
+
+
+def run(args):
+
+    path = args.get("--output")
+    if path is None:
+        path = PARTS_FILE
+
+    index = args.get('--index')
+    if index is not None:
+        if "://" not in index:
+            index = "%s%s" % ("file://", os.path.join(os.getcwd(), index))
+        output = urllib.request.urlopen(index).read()
+    else:
+        # XXX: fetch the index from the wiki
+        output = "{}"
+
+    master_parts_list = _process_index(output)
+
+
+    _write_gzipped_parts_list(path, master_parts_list)
+
+
+    if args['--debug']:
+        print(yaml.dump(master_parts_list, default_flow_style=False))
+
+    return args
+
+def is_valid_parts_list(parts_list, parts):
+    for partname in parts:
+        if partname not in parts_list.keys():
+            logging.error(
+                "Part '%s' is referenced by another part but not included in the master parts list" % partname)
+            return False
+
+    return True
+
+def _write_gzipped_parts_list(path, master_parts_list):
+    logging.debug("Writing gzipped parts list to %s", path)
+    with gzip.open(path, "w") as fgp:
+        fgp.write(yaml.dump(master_parts_list,
+                            default_flow_style=False).encode("utf-8"))
+
+
+def _namespaced_partname(key, partname):
```

Comment:
@elopio will probably mention the PEP about using a `get` in front of this method.

fwiw, I'd like this moved to right above the list comprehension that uses it for easier reading.